Sources/MusicModel/Documentation/Design.md

```diff
@@ -66,4 +66,182 @@ Identifiers for each event are stored in a `Dictionary<VoiceID,IntervalSearchTre
+
+Ultimately, though, there is no more type-safety than when declaring all types as `Any`, as Swift has no concept analogous to Scala's [sealed traits](https://underscore.io/blog/posts/2015/06/02/everything-about-sealed.html). As we will see in the next section, this may not desirable, anyway.
+
+### Most Safe
+
+On the other side of the spectrum from the type-unsafe world of `Any` and conditional downcasting are enums.
+
+```Swift
+enum MusicalItem {
+    case pitch(Pitch)
+    case dynamic(Dynamic)
+    case articulation(Articulation)
+}
+```
+
+The event described above could then be re-represented as:
+
+```Swift
+let event: [MusicalItem] = [.pitch(Pitch(60)), .dynamic(.fff), .articulation(.staccato)]
+```
+
+Upon representing the event, the Swift compiler enforces that we handle each case:
+
+```Swift
+for item in event {
+    switch item {
+    case .pitch(let pitch): ...
+    case .dynamic(let dynamic): ...
+    case .articulation(let articulation): ...
+    }
+}
+```
+
+If the composer deems the `FireCanon` type critical to their musical practice, a case can be added to `MusicalItem`:
+
+```Swift
+enum MusicalItem {
+    case pitch(Pitch)
+    case dynamic(Dynamic)
+    case articulation(Articulation)
+    case fireCannon(FireCannon)
+}
+```
+
+A downstream renderer would not compile because it does not handle all specified cases.
+
+```Swift
+for item in event {
+    switch item {
+    case .pitch(let pitch): ...
+    case .dynamic(let dynamic): ...
+    case .articulation(let articulation): ...
+    // error: Switch must be exhaustive
+    }
+}
+```
+
+### Final Tagless Style
+
+There are two goals for this abstract model of musical information: that it is type-safe, and arbitrarily extensible by developers downstream.
+
+An interesting avenue for research into solving this problem (the [expression problem](https://en.wikipedia.org/wiki/Expression_problem)) is the use of the "final tagless style". The final tagless style in Swift, uses static, `Self`-returning, methods required by composable protocols instead of enums or class hierarchies.
+
+This is a compelling approach because protocol requirements are enforced by the compiler (satisfying our type-safety goal) and cases can be extended arbitrarily (satistying our exstensibility goal).
+
+Let's try to build up our `MusicalItem` value this way:
+
+```Swift
+protocol MusicalItem {
+    static func pitch(_: Pitch) -> Self
+    static func dynamic(_: Dynamic) -> Self
+    static func articulation(_: Articulation) -> Self
+}
+```
+
+We can create a variety of concrete renderers, one of which could be:
+
+```Swift
+struct Renderer: MusicalItem {
+    static func pitch(_ pitch: Pitch) -> Renderer { ... }
+    static func articulation(_ articulation: Articulation) -> Renderer { ... }
+    static func dynamic(_ dynamic: Dynamic) -> Renderer { ... }
+    static func collection([Self]) -> Renderer { ... }
+}
+```
+
+> There is still some mystery about the actual implementation of `Renderer` ...
```

Comment:
If `Renderer` is to take the place of e.g. a `switch` statement, it needs to do some kind of 'pattern matching'. What about if we have some `PitchRenderer` which implements every function but `.pitch` in some trivial way and then does the actual thing we want it to do in the `.pitch` function? A nice way to cope with the 'trivial' case might be to have each `static func` return a `Renderer?`, a.k. a `Self?` in the protocol.

When you actually have say a drawing function, it may have a non-trivial pass through every function (non-nil return value).